jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeNode.kt

```diff
@@ -77,68 +93,118 @@ class S3TreeDirectoryNode(private val bucket: S3VirtualBucket, parent: S3LazyLoa
-                ?.map { S3TreeObjectVersionNode(bucket, this, key, it.size(), it.lastModified(), it.versionId()) as S3TreeNode }
-                ?: emptyList()
-
-            if (responseIterator?.hasNext() == true) {
-                return nextPage + S3TreeContinuationNode(
-                    bucketName,
-                    this,
-                    "${this.key}/${message("s3.load_more")}",
-                    (nextPage.last() as S3TreeObjectVersionNode).versionId
-                )
+            val response = runBlocking {
+                bucket.listObjectVersions(key, continuationMarker?.keyMarker, continuationMarker?.versionId)
+            }
+
+            return mutableListOf<S3TreeNode>().apply {
+                response?.versions()
+                    ?.filter { it.key() == key && it.versionId() != NOT_VERSIONED_VERSION_ID }
+                    ?.map { S3TreeObjectVersionNode(this@S3TreeObjectNode, it.versionId(), it.size(), it.lastModified()) }
+                    ?.onEach { add(it) }
+
+                if (response?.isTruncated == true) {
+                    val nextKey = response.nextKeyMarker()
+                    val nextVersion = response.nextVersionIdMarker()
+
+                    add(
+                        S3TreeContinuationNode(
+                            bucket,
+                            this@S3TreeObjectNode,
+                            this@S3TreeObjectNode.key,
+                            VersionContinuationToken(nextKey, nextVersion)
+                        )
+                    )
+                }
             }
-            return nextPage
         }
+
         return emptyList()
     }
 }
 
-class S3TreeObjectVersionNode(bucket: S3VirtualBucket, parent: S3TreeObjectNode, key: String, size: Long, lastModified: Instant, val versionId: String) :
-    S3TreeObjectNode(bucket, parent, key, size, lastModified) {
+class S3TreeObjectVersionNode(parent: S3TreeObjectNode, override val versionId: String, override val size: Long, override val lastModified: Instant) :
+    S3TreeNode(parent.bucket, parent, parent.key), S3Object {
 
-    override fun getName(): String {
-        // For not versioned buckets api can return versionId as literal 'null' so we avoid propagating null string to UI.
-        val versionId = if (versionId != NOT_VERSIONED_VERSION_ID) versionId else (parent as S3TreeObjectNode).name
-        val originalExtension = FileUtilRt.getExtension((parent as S3TreeObjectNode).name)
-        val extension = if (originalExtension.isNotBlank()) ".$originalExtension" else ""
+    init {
+        icon = parent.icon
+    }
 
-        return "$versionId$extension"
+    override fun directoryPath(): String = (parent as S3TreeObjectNode).directoryPath()
+
+    override fun fileName(): String {
+        val parentObjectName = (parent as S3TreeObjectNode).fileName()
+
+        val filenamePrefix = FileUtilRt.getNameWithoutExtension(parentObjectName) + "@" + versionId
```

Comment:
i think `@` is valid on all modern file systems so hopefully we won't have any issues here

Reply:
I checked windows reserved chars (and tests passed on linux, osx, and windows) :D